iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/gcp/ExcessivePermissionsCheck.java

```diff
@@ -0,0 +1,90 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.terraform.checks.gcp;
+
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.iac.common.api.checks.SecondaryLocation;
+import org.sonar.iac.common.checks.TextUtils;
+import org.sonar.iac.terraform.api.tree.ExpressionTree;
+import org.sonar.iac.terraform.checks.AbstractNewResourceCheck;
+import org.sonar.iac.terraform.symbols.ListSymbol;
+
+@Rule(key = "S6406")
+public class ExcessivePermissionsCheck extends AbstractNewResourceCheck {
+
+  private static final String MESSAGE = "This role grants more than %d sensitive permissions. Make sure they are all required.";
+  private static final String SECONDARY_MESSAGE = "Sensitive permission";
+  public static final int DEFAULT = 5;
+
+  private static final Set<String> SENSITIVE_FEATURE_SUFFIXES = Set.of("abort", "access", "add", "allocate", "analyze", "apply", "approve", "associate", "attach", "begin",
+      "bind", "call", "cancel", "clear", "close", "compute", "connect", "create", "delete", "deploy", "destroy", "detach",
+      "disable", "drop", "enable", "evict", "exec", "import", "install", "invoke", "listVulnerabilities", "manage",
+      "migrate", "move", "mutate", "patch", "pause", "proxy", "publish", "purchase", "purge", "put", "reject", "remove",
+      "reopen", "replace", "rerun", "reset", "resize", "restart", "restore", "resume", "rollback", "rotate", "run",
+      "sample", "scan", "send", "set", "sign", "sourceCodeGet", "sourceCodeSet", "start", "stop", "suspend", "undelete",
+      "undeploy", "update", "upload", "use", "validate", "write");
+
+  private static final Set<String> SENSITIVE_FEATURE_SUFFIX_ELEMENTS = Set.of("login", "create", "delete", "set");
+
+  @RuleProperty(
+    key = "max",
+    defaultValue = "" + DEFAULT)
+  public int max = DEFAULT;
+
+  @Override
+  protected void registerResourceConsumer() {
+    register(Set.of("google_organization_iam_custom_role", "google_project_iam_custom_role"),
+      resource -> {
+        ListSymbol permissions = resource.list("permissions");
+        List<ExpressionTree> sensitivePermissions = resource.list("permissions")
+          .getItemIf(isSensitivePermission())
+          .collect(Collectors.toList());
```

Comment:
Correct me if I am wrong: we are aiming for a future where we won't have to juggle between `Resource`s and `Tree`s (like in this case) but do everything on `Resource`s?

Reply:
I don't get the questions. Can you provide more context regarding these lines?

Reply:
Now I see I did not spot something:
On line 61: replace `resource.list("permissions")` with `permissions` ? (as you've already got it on line 60?)

Answer to your question:
On line 60 you're getting the `Resource` and on line 61 - `List<ExpressionTree>`. Are we aiming for future where we will be getting something like `List<ItemSumbols>` on line 61, so that we don't have to deal both with the `Resource`(s) and `Tree`(s) class hierarchy in a Check class ? (Hint 'No' is a legitimate answer)

Reply:
Thank you for the clarification. Yes, I think that later we could also have symbols for list items. And we can get away from Tree/AST in the medium term. For now, this is an in-between solution and a baby step to switch from trees to symbols.